src/box/box.cc

```diff
@@ -1805,6 +1805,218 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+	/** Syncs to wait for. */
+	uint64_t syncs[VCLOCK_MAX];
+	/**
+	 * A bitmap holding replica ids whose vclocks were already collected.
+	 */
+	vclock_map_t collected_vclock_map;
+	/** The fiber waiting for vclock. */
+	struct fiber *waiter;
+	/** Collected vclock. */
+	struct vclock *vclock;
+	/** The request deadline. */
+	double deadline;
+	/** Whether the request is timed out. */
+	bool is_timed_out;
+};
+
+/**
+ * A trigger executed on each ack to collect up to date remote node vclocks.
+ * When an ack comes with requested sync for some replica id, ack vclock is
+ * accounted.
+ */
+static int
+check_sync_on_ack(struct trigger *trigger, void *event)
+{
+	struct replication_ack *ack = (struct replication_ack *)event;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = ack->source;
+	uint64_t sync = data->syncs[id];
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->sync >= sync) {
+		vclock_max_ignore0(data->vclock, ack->vclock);
+		bit_set(&data->collected_vclock_map, id);
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** A trigger querying relay's next sync value once it becomes operational. */
+static int
+relay_get_sync_on_start(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)event;
+	if (replica->anon)
+		return 0;
+	struct relay *relay = replica->relay;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = replica->id;
+	/* Already accounted. */
+	if (bit_test(&data->collected_vclock_map, id))
+		return 0;
+	if (relay_next_sync(relay, &data->syncs[id], data->deadline) != 0) {
+		diag_clear(diag_get());
+		data->is_timed_out = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** Find the minimal vclock which has all the data confirmed on a quorum. */
+static int
+box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
+{
+	/*
+	 * How many vclocks we should see to be sure that at least one of them
+	 * contains all data present on any real quorum.
+	 */
+	int vclock_count = MAX(1, replicaset.registered_count) -
+			   replication_synchro_quorum + 1;
+	/*
+	 * We should check the vclock on self plus vclock_count - 1 remote
+	 * instances.
+	 */
+	vclock_copy(confirmed_vclock, &replicaset.vclock);
+	if (vclock_count <= 1)
+		return 0;
+
+	struct sync_trigger_data data = {
+		.syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.is_timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
```

Comment:
We are trying to get latest data. Getting data works via appliers. I suppose we use relays here only because they can do the pings, and appliers can only send ACKs, but not request them? Which means the feature works only if the instance, doing the linearizable txn, is also a replication source on at least quorum of nodes, right? Essentially, it can't be used on some rogue (maybe anon) replica to get latest data if the replica itself is not a part of the quorum. Might be worth reflecting in the documentation. But up to you.

Reply:
Good point, I'll write a couple of words about that in the doc request:

```
    Note: starting a linearizable transaction requires that the node is the
    replication **source** for at least N - Q + 1 remote replicas. Here `N` is
    the count of registered nodes in the cluster and `Q` is
    `replication_synchro_quorum` value (the same as
    `box.info.synchro.quorum`). This is the implementation limitation. For
    example, you may start linearizable transactions on any node of a
    cluster in full-mesh topology, but you can't perform linearizable
    transactions on anonymous replicas, because noone replicates **from**
    them.

```

Reply:
Hm, TBH this scheme looks quite unnatural. The channel is reversed. Did you think about trying to implement mutual syncs for applier<->relay pair? So as not only relay would send syncs, but also applier could send them at least on demand? Maybe it wouldn't be too complicated?

In future it could also be of use in general too. Applier, by getting vclock of the remote replication source, could estimate and report how big is the gap and reflect it in `box.info.replication`.

Reply:
What is the progress here? Was it decided to leave it upside-down?

Reply:
I'd like to leave it upside down for now. We can always return to that later. I filed an [issue](https://github.com/tarantool/tarantool/issues/7723).

The protocol remains compatible: both relay and applier may encode a vclock and a vclock_sync in a heartbeat.
Currently relay only encodes vclock_sync, and applier encodes both fields. When and if we decide to implement the reverse channel, relay may start sending vclock in the heartbeat. Both vclock and vclock_sync are fields in the heartbeat body, so everything stays compatible.

I don't want to do that now because it would require making relay thread poll tx to find out the latest vclock. This is against relay architecture: it doesn't perform requests to tx thread.